pkg/util/pki/othername_sans.go

```diff
@@ -0,0 +1,89 @@
+/*
+Copyright 2020 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pki
+
+import (
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	v1 "github.com/cert-manager/cert-manager/pkg/apis/certmanager/v1"
+)
+
+const tagOtherName = 0
+
+var oidExtensionSubjectAltName = asn1.ObjectIdentifier{2, 5, 29, 17}
+
+type otherName struct {
+	TypeID asn1.ObjectIdentifier
+	Value  any
+}
+
+// StringValueLikeType type for asn1 encoding. This will hold
+// our utf-8 encoded string.
+type StringValueLikeType struct {
+	A string `asn1:"utf8"`
+}
+
+func buildExtraExtensionsOtherNameSANsForCertificate(crt *v1.Certificate) ([]pkix.Extension, error) {
+	ret := make([]pkix.Extension, 0)
+
+	for index, otherNameSAN := range crt.Spec.OtherNameSANs {
+		oidRegex, err := regexp.Compile("oid:(([[:digit:]]+|\\.)+)")
+		if err != nil {
+			return []pkix.Extension{}, err
+		}
+
+		output := oidRegex.FindStringSubmatch(otherNameSAN.Type)
+		if len(output) <= 1 {
+			return []pkix.Extension{}, fmt.Errorf("Invalid OID in otherName SAN supplied: index: %d, type %s", index, otherNameSAN.Type)
+		}
+
+		// extract first match of first capture group
+		OIDString := output[1]
+
+		parsedOID := asn1.ObjectIdentifier{}
+		// we know this is a "valid" OID so convert:
+		for _, elem := range strings.Split(OIDString, ".") {
+			oidElem, err := strconv.Atoi(elem)
+			if err != nil {
+				return []pkix.Extension{}, fmt.Errorf("Unexpected error while parsing otherName OID SAN: index: %d, type %s, error: %s", index, otherNameSAN.Type, err.Error())
+			}
+			parsedOID = append(parsedOID, oidElem)
+		}
+
+		asn1otherNameSAN, err := asn1.MarshalWithParams(otherName{
```

Comment:
We should wrap this in another struct (see https://samfira.com/post/golang-x-509-certificates-and-othername/#:~:text=//%20GeneralNames%20type%20for%20asn1%20encoding%0Atype%20GeneralNames%20struct%20%7B%0A%20%20%20%20OtherName%20OtherName%20%60asn1%3A%22tag%3A0%22%60%0A%7D)